Kudu.Core/Functions/FunctionManager.cs

```diff
@@ -49,52 +52,291 @@ public FunctionManager(IEnvironment environment, ITraceFactory traceFactory)
+        {
+            var functionPath = GetFunctionPath(functionName);
+            var functionFiles = FileSystemHelpers.GetFiles(functionPath, "*.*", SearchOption.TopDirectoryOnly)
+                .Where(p => Path.GetFileName(p).ToLowerInvariant() != "function.json").ToArray();
+
+            if (functionFiles.Length == 0)
+            {
+                return functionPath;
+            }
+            else if (functionFiles.Length == 1)
+            {
+                // if there is only a single file, that file is primary
+                return functionFiles[0];
+            }
+            else
+            {
+                // if there is a "run" file, that file is primary
+                string functionPrimary = null;
+                functionPrimary = functionFiles.FirstOrDefault(p => Path.GetFileNameWithoutExtension(p).ToLowerInvariant() == "run");
+                if (string.IsNullOrEmpty(functionPrimary))
+                {
+                    // for Node, any index.js file is primary
+                    functionPrimary = functionFiles.FirstOrDefault(p => Path.GetFileName(p).ToLowerInvariant() == "index.js");
+                    if (string.IsNullOrEmpty(functionPrimary))
+                    {
+                        // finally, if there is an explicit primary file indicated
+                        // in config, use it
+                        JToken token = functionInfo["source"];
+                        if (token != null)
+                        {
+                            string sourceFileName = (string)token;
+                            functionPrimary = Path.Combine(functionPath, sourceFileName);
+                        }
+                    }
+                }
+
+                if (string.IsNullOrEmpty(functionPrimary))
+                {
+                    // TODO: should this be an error?
+                    return functionPath;
+                }
+                return functionPrimary;
+            }
+        }
+
+        private Uri FilePathToVfsUri(string filePath, bool isDirectory = false)
+        {
+            filePath = filePath.Substring(_environment.RootPath.Length).Trim('\\').Replace("\\", "/");
+            return new Uri($"{_environment.AppBaseUrlPrefix}/api/vfs/{filePath}{(isDirectory ? "/" : string.Empty)}");
+        }
+
+        private Uri GetFunctionHref(string functionName)
+        {
+            return new Uri($"{_environment.AppBaseUrlPrefix}/api/functions/{functionName}");
+        }
+
+        private string GetFunctionPath(string name)
+        {
+            var path = Path.Combine(_environment.FunctionsPath, name);
+            if (FileSystemHelpers.DirectoryExists(path))
+            {
+                return path;
+            }
+
+            throw new FileNotFoundException($"Function ({path}) does not exist");
+        }
+
+        private static async Task<IEnumerable<GitHubContent>> GetTemplatesFromGithub()
+        {
+            using (var client = GetHttpClient())
```

Comment:
Can't this client be reused for the lifetime of this class? It's properties don't appear to change.